### DISCUSSION TITLE
Per-project /public directory adjustment?

### DISCUSSION THREAD

**Discussion by marcamos** [beyondcode/herd-community#1207] [Q&A] [answered]
I feel like I'm 75% of the way to figuring this out, but I've hit a wall with the few things I've tried. In short, a number of my projects are setup as `/project-name/public` and Herd serves them perfectly. A couple, though, are not. For example, one of my projects is setup to serve the public from `/another-project/demo` … and I can't figure out how to get Herd to serve **_this one project_** from `/demo`?

Searching around, I see the concept of custom drivers, but I blank on what to write in it, where to put it, or even why it's  so much harder than it was in Valet.

Any tips? Thank you!

**Comment by marcamos** [comment, marked as answer]
I've sorted it out with help from Sebastian from BeyondCode! Here's exactly what I did:

1. In `~/Library/Application Support/Herd/config/valet/` I created a new directory called `Drivers` (as it did not yet exist).
2. In `Drivers` I added a new file called `DemoValetDriver.php` (I chose "Demo" at the beginning of the file-name because I'll have a few projects that are demos and they'll have this same requirement). **It's important that the file-name _end_ in `ValetDriver.php`**.
3. I copied/pasted the [documentation's example driver code](https://herd.laravel.com/docs/macos/extending-herd/custom-drivers#custom-laravel-driver-example) into `DemoValetDriver.php` and made the following adjustments:
   * Added `<?php` to the top of the file.
   * Changed `CustomLaravelValetDriver` on line `7` to `DemoValetDriver`.
   * Changed all instances of `/web` to `/demo`.
   * Made sure all instances of `/demo/index.php` were changed to `/demo/index.html`, as my project is a VueJS project and doesn't rely on PHP at all. **You may not need to make this change.**

And that's it!

**Reply by InforPsico** [reply]
Hi, I'm having the same problem you had, but I can't figure out your solution. 

I have followed the route indicated in the documentation for Windows: `%USERPROFILE%\.config\herd\config\valet\Drivers`

I've created a Drivers folder and inside it, I've created a DemoValetDriver.php file with the code shown here.

My goal is to get Herd to recognize index.php within public_html, so I've replaced all references to "web" with "public_html."

Unfortunately, I keep getting a 404. What am I doing wrong?

```
namespace Valet\Drivers\Custom;

use Valet\Drivers\LaravelValetDriver;

class DemoValetDriver extends LaravelValetDriver
{
    /**
     * Determine if the driver serves the request.
     */
    public function serves(string $sitePath, string $siteName, string $uri): bool
    {
        return file_exists($sitePath.'/public_html/index.php') &&
               file_exists($sitePath.'/artisan');
    }

    /**
     * Determine if the incoming request is for a static file.
     */
    public function isStaticFile(string $sitePath, string $siteName, string $uri)/*: string|false */
    {
        if (file_exists($staticFilePath = $sitePath.'/public_html'.$uri)
           && is_file($staticFilePath)) {
            return $staticFilePath;
        }

        $storageUri = $uri;

        if (strpos($uri, '/storage/') === 0) {
            $storageUri = substr($uri, 8);
        }

        if ($this->isActualFile($storagePath = $sitePath.'/storage/app/public'.$storageUri)) {
            return $storagePath;
        }

        return false;
    }

    /**
     * Get the fully resolved path to the application's front controller.
     */
    public function frontControllerPath(string $sitePath, string $siteName, string $uri): ?string
    {
        if (file_exists($staticFilePath = $sitePath.'/public_html'.$uri)
           && $this->isActualFile($staticFilePath)) {
            return $staticFilePath;
        }

        return $sitePath.'/public_html/index.php';
    }

        /**
     * Get the logs paths for the application to show in Herds log viewer.
     */
    public function logFilesPaths() {
        return ["/storage/logs"];
    }

    /**
     * Display information about the application in the information tab of the Sites UI.
     * For Laravel, it's the output of the `php artisan about` command.
     */
    public function siteInformation(string $sitePath, string $phpBinary): array
    {
        try {
            $process = new Process([
                $phpBinary,
                'artisan',
                'about',
                '--json'
            ], $sitePath);

            $process->mustRun();

            $result = json_decode($process->getOutput(), true);
        } catch (\Exception $e) {
            $result = [];
        }

        return [
            ...$result,
        ];
    }
}
```

**Reply by marcamos** [reply]
@InforPsico The only thing I don't see in your file is:

> Added `<?php` to the top of the file.

Might be that?

**Reply by InforPsico** [reply]
No thanks. That's a given. I can confirm that's not the problem. The question is, have you managed to modify it to detect folders other than public?

**Reply by marcamos** [reply]
I've been serving from `/demo/` thanks to the adjustments I wrote-up above.
